### ISSUE TITLE
Regular ACK frames when multipath has been negotiated

### ISSUE THREAD

**Issue by michael-eriksson** [quicwg/multipath#181]
# Introduction

It should be allowed to send regular ACK frames also after multipath has been negotiated.

# Reasoning

Sending acknowledgments over other paths than where the packet was received is complex, which is recognised in the Packet Scheduling section of the draft. A straight-forward implementation may therefore never send cross-path acknowledgments. Allowing it to use regular ACK frames would reduce the special-case code for multipath mode and also save at least one byte on every acknowledgment.

The specification should be updated to (explicitly) say that an endpoint MAY send regular ACK frames also after multipath has been negotiated.

**Comment by qdeconinck** [comment]
Sending regular ACK frames is still allowed after Multipath has been negotiated, but these refer to the Path ID 0.

If I understand well, the proposal is to allow considering the reception of an ACK frame over a given 4-tuple as the acknowledgment of packets sent over that 4-tuple. ACK/ACK_MP frames in multipath acknowledge packets using a specific DCID. The issue here is that there is a one-to-many relation between a 4-tuple and the Path ID (i.e., the sequence number of the DCID used). Relying on the 4-tuple to infer the DCID that may lead to acknowledgment ambiguities here.

Consider for the example the case where a client decides to rotate the DCID it uses over a given path every 2 packets (we could also consider a timer-based change where the DCID change occurs in the middle of a sending burst). It then sends over that path DCID 1 PN 0, DCID 1 PN 1, DCID 2 PN 0 and DCID 2 PN 1. What should the server sends here? If it relies on the largest DCID sequence number seen, it may send ACK(0-1), but from the client perspective, such indication is unclear (especially if, e.g., one of the packets sent with DCID 1 has been lost).

**Comment by yfmascgy** [comment]
> # Introduction
> It should be allowed to send regular ACK frames also after multipath has been negotiated.
> 
> # Reasoning
> Sending acknowledgments over other paths than where the packet was received is complex, which is recognised in the Packet Scheduling section of the draft. A straight-forward implementation may therefore never send cross-path acknowledgments. Allowing it to use regular ACK frames would reduce the special-case code for multipath mode and also save at least one byte on every acknowledgment.
> 
> The specification should be updated to (explicitly) say that an endpoint MAY send regular ACK frames also after multipath has been negotiated.

The ACK_MP's path ID is tied to packet number space. Removing it will introduce ambiguity. I also feel the benefit of saving one byte is not very significant.

**Comment by michael-eriksson** [comment]
@qdeconinck, I don't think that the mechanism for path CID changes has been defined, has it? I, for one, would *really* like to have stable path identifiers (that are identical in both directions).

I'm not sure I understand what you are suggesting above, please correct me if I'm wrong below.

It seems like you suggest that the packet numbers are reset for the path when the destination connection identifier changes, is that correct? That would imply that you would then need a new instance of the loss detection (which is packet number based) when the CID changes, including restarting RTT estimates etc. Maybe even worse, you would have double loss detection instances for the path until all packets sent with the previous CID are acknowledged or declared lost. Is this correct?

**Comment by michael-eriksson** [comment]
> The ACK_MP's path ID is tied to packet number space. Removing it will introduce ambiguity. I also feel the benefit of saving one byte is not very significant.

The implicit path id of the regular ACK frame would be the path id over which is was transmitted. A receiver must anyway always be able to understand on which path a packet arrives, so it can set the Path Identifier field of the ACK_MP frame to the correct value.

The main point with this issue is not to save one or a few bytes (but see below). What I am after is to reduce the special case code for multipath compared to regular unipath QUIC when multipath isn't enabled. That leads to simpler and cleaner code, which is easier to test and maintain.

When it comes to saving bytes, that might seem silly in the times of terabit fibers. However, the world isn't that simple...
- ACK-only packets are not congestion controlled, and keeping them small is good for the stability of the Internet.
- Some paths are very asymmetric and can even be ACK limited; smaller ACKs is a good thing here as well.
- My employer, Ericsson, produce cellular infrastructure. The uplink, from the mobile device to the base station, is often a limitation. For media streaming, which is like 70+% of the mobile broadband traffic, almost only ACKs go on the uplink. Saving a completely redundant byte here can actually be relevant, especially if most of the header (including the connection identifier) is compressed away.

**Comment by qdeconinck** [comment]
> @qdeconinck, I don't think that the mechanism for path CID changes has been defined, has it? I, for one, would really like to have stable path identifiers (that are identical in both directions).

Actually, this is somewhat implicitly discussed in [Section 4.3.3](https://quicwg.org/multipath/draft-ietf-quic-multipath.html#section-4.3.3). Over a given path (4-tuple), an endpoint can change the DCID it uses to send packets. The sequence number of the DCID used by the client over a given path may be different from the server one, yes.

> I'm not sure I understand what you are suggesting above, please correct me if I'm wrong below.

> It seems like you suggest that the packet numbers are reset for the path when the destination connection identifier changes, is that correct?

During some time, the sender will have two space numbers associated to that path, indeed (the one of the old DCID seq num and the new one). At most, you may need to maintain max_active_cid_limit sending application data spaces. The receiver can also track the 4-tuple on which it observes the CID to bind its ACK_MP there.

> That would imply that you would then need a new instance of the loss detection (which is packet number based) when the CID changes, including restarting RTT estimates etc. Maybe even worse, you would have double loss detection instances for the path until all packets sent with the previous CID are acknowledged or declared lost. Is this correct?

I'm not sure the recovery structure needs to be duplicated though. Given you are using the same 4-tuple (just rotating the CID), you can keep the same RTT. You can also update your recovery structure to keep track of the space identifier of the sent packets. Such complexity will be located at the sender-side.

**Comment by mirjak** [comment]
@qdeconinck I wasn't assuming that you would reset the packet number when only the CID changes; this is not a new path. Why is that needed?

I guess this is a separate issue we need to discuss and clarify.

**Comment by qdeconinck** [comment]
> @qdeconinck I wasn't assuming that you would reset the packet number when only the CID changes; this is not a new path. Why is that needed?

Because the 4-tuple perceived by the receiver may change while the one used by the sender might not. The CID sequence number provides an unambiguous matching to a packet number space for both the sender and the receiver without decrypting the packet.

**Comment by qdeconinck** [comment]
At sender side, yes. But at receiver side, this is trickier, especially if the packet number encoding truncates the actual packet number (which packet number should you expect from a new CID to derive the actual one?).

**Comment by huitema** [comment]
To understand the relation between CID and number space, consider the case of a path on which 1 + 2^32 have been sent using CID number X. Now, the sender decides to refresh the CID to number Y. What should be the PN for the next packet? If you said "2+2^32", you are going to have a bug, because the packet header can use at most 32 bits to encode the packet. The packet will look like:

| 1rtt (PN length=4) | CID (seq num=Y) | 32 bit PN=0x00000002 | payload... |
|-------------|-------------------------|--------------|----------|

The receiver of the packet is suppose to expand the PN number to 64 bits to get the IV used by AEAD. To do that, it looks at the CID, find the corresponding number space, and infer the upper 32 bits from past numbers. If the CID was X, the past number would be "1+2^32", and the PB would expand to "0x0000000100000002". But that it is a first use of CID Y, and the receiver has not seen any previous path ID. So it expands the number to ""0x0000000000000002". Both sender and receiver have to use that value, otherwise decryption will fail.

Bottom line: number spaces are really tied to CID numbers.

**Comment by michael-eriksson** [comment]
> Bottom line: number spaces are really tied to CID [sequence] numbers.

That is not true.

To be able to decode an incoming packet, the receiving endpoint must be able to map the CID to the path, and in particular get two pieces of path-specific information:
- The `largest_pn` state variable to be able to decode (i.e., expand) the truncated packet number
- The path identifier to create the nonce for the AEAD decryption.

This information can be for instance be stored in a hash map that has the CID as key. The CID sequence number is not relevant, and in particular the path identifier can (and should) be different from the CID sequence number.

**Comment by yfmascgy** [comment]
> A straight-forward implementation may therefore never send cross-path acknowledgments. Allowing it to use regular ACK frames would reduce the special-case code for multipath mode and also save at least one byte on every acknowledgment.

Completely disabling cross-path acknowledgement will introduce new issues and makes communication fragile. For example, consider the case a sender has just retired the CID of a path and has no available CIDs to use on that path anymore, but it later may still receive a couple of packets from the peer on that path. If we completely disable cross-path acknowledgement, then these later received packets cannot be acknowledged. Using ACK_MP, this is not a problem as it can be sent from other available paths, which makes the communication more robust.

**Comment by mirjak** [comment]
I don't think we need any further normative language in the draft. It's allowed to use v1 ACK frames. This is what the draft says:

> As soon as the negotiation of multipath support is completed, endpoints SHOULD use ACK_MP frames instead of ACK frames to acknowledge application data packets, including 0-RTT packets, received on path ID 0 after the handshake concluded.

> If for any reason ACK frames are received in 1-RTT packets while the state of multipath negotiation is ambiguous, they MUST be interpreted as acknowledging packets sent on path 0.

**Comment by mirjak** [comment]
Resetting the packet number whenever a new CID is used it is deviation from RFC 9000. I opened issue #182 for this.

**Comment by michael-eriksson** [comment]
> > A straight-forward implementation may therefore never send cross-path acknowledgments. Allowing it to use regular ACK frames would reduce the special-case code for multipath mode and also save at least one byte on every acknowledgment.
> 
> Completely disabling cross-path acknowledgement will introduce new issues and makes communication fragile. For example, consider the case a sender has just retired the CID of a path and has no available CIDs to use on that path anymore, but it later may still receive a couple of packets from the peer on that path. If we completely disable cross-path acknowledgement, then these later received packets cannot be acknowledged. Using ACK_MP, this is not a problem as it can be sent from other available paths, which makes the communication more robust.

In general, I think that there are cases when sending an `ACK_MP` frame is both elegant and useful. However, these events are likely very rare. They are probably also rather specific and therefore hard to detect in time with generic algorithms.

A straight-forward implementation would just recognise the path-specific problems and quickly retransmit on some other path, possibly with some overhead because of redundant transmission. A simple implementation could also have *some* special-case code for path problems, the key point with this issue is that steady state transmission should be as similar to unipath QUIC as possible.

When it comes to your case above, it looks a bit odd to be honest. Why would the "sender" retire the CID if there's a chance that it will receive more packets on that path? Isn't path closing supposed to happen in an ordered fashion with `PATH_ABANDON` frames and only after that is the CID retired?

**Comment by michael-eriksson** [comment]
> I don't think we need any further normative language in the draft. It's allowed to use v1 ACK frames. This is what the draft says:
> 
> > As soon as the negotiation of multipath support is completed, endpoints SHOULD use ACK_MP frames instead of ACK frames to acknowledge application data packets, including 0-RTT packets, received on path ID 0 after the handshake concluded.
> 
> > If for any reason ACK frames are received in 1-RTT packets while the state of multipath negotiation is ambiguous, they MUST be interpreted as acknowledging packets sent on path 0.

The point here is to develop and specify the best possible protocol which can also lead to nice, clean and efficient implementations with minimal special-case code for multipath.

With this issue, I wanted to slightly adjust the semantics of the `ACK` frame when multipath is enabled; the currently defined semantics are pretty (but not fully) useless. As indicated by the reasons stated above, there are more useful semantics. An implementation would still be allowed to send `ACK_MP` frames, so there is no reduction of existing functionality.

**Comment by huitema** [comment]
Anything that relies on 4-tuple identification leads to ambiguity, because of NATs. It is much more robust to always use ACK_MP and always specify the CID number, even if that costs 1 extra byte. Specifying that naked ACK always maps to CID 0 achieves that.

**Comment by huitema** [comment]
Also, ambiguity in ACK processing will result in failures in case of NAT rebinding. The peer receives an ACK for "this path", but because of rebinding "this path" is in fact a new path, with a new number space starting at 0. The ACK for "this path" reference packet numbers that have not yet been sent, which is a protocol violation, causing the termination of the connection.
